hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/marker/MarkerCreationDispatchingRunnable.java

```diff
@@ -66,8 +66,9 @@ public void run() {
 
     // Only fetch pending marker creation requests that can be processed,
     // i.e., that markers can be written to a underlying file
-    for (String markerDir : markerDirStateMap.keySet()) {
-      MarkerDirState markerDirState = markerDirStateMap.get(markerDir);
+    for (Map.Entry<String, MarkerDirState> entry : markerDirStateMap.entrySet()) {
+      String markerDir = entry.getKey();
+      MarkerDirState markerDirState = entry.getValue();
```

Comment:
`markerDirState.fetchPendingMarkerCreationRequests` seems already thread safe ?

Reply:
but markerDirStateMap is not thread safe. request add/delete MarkerDirState in markerDirStateMap, and MarkerCreationDispatching traversal markerDirStateMap in other thread. use hashMap may cause ConcurrentModificationException, and ConcurrentHashMap can fix it.

Reply:
> `markerDirState.fetchPendingMarkerCreationRequests` seems already thread safe ?

but markerDirStateMap is not thread safe. request add/delete MarkerDirState in markerDirStateMap, and MarkerCreationDispatching traversal markerDirStateMap in other thread. use hashMap may cause ConcurrentModificationException, and ConcurrentHashMap can fix it.

Reply:
Got it, maybe we can add more document about the details in the code.

Reply:
> Got it, maybe we can add more document about the details in the code.

ok, I will add more document later.

Reply:
+1.  @novisfff could you use HUDI-4574 which is the original ticket for the timeline-server-based marker issue?  I was investigating the issue and identified the same thread-safe problem.

Reply:
> +1. @novisfff could you use [HUDI-4574](https://issues.apache.org/jira/browse/HUDI-4574) which is the original ticket for the timeline-server-based marker issue? I was investigating the issue and identified the same thread-safe problem.

done.